feat(docs): Add power profiler

Nicell opened the pull request:
This adds a battery life estimator to the docs website. This is currently working quite well I'd say, but I'd like to get some feedback before I go into the details of cleaning up the code. Try it here: https://deploy-preview-312--zmk.netlify.app/power-profiler

Todo:

- [ ] Document how the calculations are done and some of the methodology
- [ ] Document how to add new boards or features
- [x] Add MIT license headers (as opposed to the rest of the docs)

innovaker commented:
@Nicell, it looks like 2871dca9bfce2588e23da0905d7142ef5ff1c1a1 has affected this PR and GitHub isn't handling it gracefully.  Please could you rebase onto `HEAD` before your next push.  Thanks!

joric commented:
I need uW/uA switch, I'm way too used to uA rather than to uW.

innovaker commented:
It's looking good @Nicell!

Technical reviews aside, my primary concern with a "ZMK-branded" power profiler will be its accuracy.  What margin of error are we talking here?  What data is it based on?  Can the estimated margins of error be represented visually?

Even with the disclaimer, having large margins of error will unfortunately lead to bad mouthing about ZMK.

Nicell commented:
> Technical reviews aside, my primary concern with a "ZMK-branded" power profiler will be its accuracy. What margin of error are we talking here? What data is it based on? Can the estimated margins of error be represented visually?
> 
> Even with the disclaimer, having large margins of error will unfortunately lead to bad mouthing about ZMK.
@innovaker 

Margin of error of the calculation should be quite small when thinking about it. Most error will come from how accurate the data is that the user inputs. Maybe we can quantify that? Especially the battery mAh.

Nicell commented:
@innovaker @joelspadin I've finished up this PR finally and would appreciate some feedback on it! For some reason netlify is failing to deploy it (says there's no changes??), so you'll have to test it locally. Thanks 😄

joelspadin commented on docs/src/data/power.js:
```diff
@@ -0,0 +1,55 @@
+    typing: 320,
+  },
+  central: {
+    idle: 500, // need to test
```
Do these still need testing?

joelspadin commented on docs/src/pages/power-profiler.js:
```diff
@@ -0,0 +1,357 @@
+          </div>
+
+          {board === "custom" && (
+            <div className="profilerSection">
```
Consider pulling this massive block out into a separate component and using something like `useContext()` to share its state back to this component.

innovaker replied:
Agreed, breaking this into subcomponents might help readability - my eyes glossed over on this file.  That said, I appreciate that'll probably increase the boilerplate code somewhat.

innovaker commented on docs/src/pages/power-profiler.js:
```diff
@@ -0,0 +1,387 @@
+
+  const [disclaimerAcknowledged, setDisclaimerAcknowledged] = useState(
+    typeof window !== "undefined"
+      ? localStorage.getItem("zmkPowerProfilerDisclaimer") === "true"
```
Nice touch 👍

innovaker commented on docs/src/pages/power-profiler.js:
```diff
@@ -0,0 +1,387 @@
+      ? {
+          powerSupply: {
+            type: psuType,
+            outputVoltage: outputV,
```
Based on other the naming conventions, presumably this is in volts rather than milli-volts or micro-volts?

innovaker commented on docs/src/pages/power-profiler.js:
```diff
@@ -0,0 +1,387 @@
+        <div className="container">
+          <h1 className="hero__title">ZMK Power Profiler</h1>
+          <p className="hero__subtitle">
+            {"Estimate your keyboard's power usage and battery life on ZMK."}
```
See above.

innovaker commented on docs/src/pages/power-profiler.js:
```diff
@@ -0,0 +1,387 @@
+    >
+      <header className={classnames("hero hero--primary", styles.heroBanner)}>
+        <div className="container">
+          <h1 className="hero__title">ZMK Power Profiler</h1>
+          <p className="hero__subtitle">
```
what does `hero__` refer tohere?

Nicell replied:
Docusaurus's class name. Reused from the home page.

innovaker commented on docs/src/pages/power-profiler.js:
```diff
@@ -0,0 +1,387 @@
+          <div className="disclaimer">
+            <h3>Disclaimer</h3>
+            <p>
+              This profiler makes many assumptions about typing activity,
+              battery characteristics, hardware behavior, and doesn&apos;t
+              account for error of user inputs. While it tries to estimate power
+              usage using real power readings of ZMK, every person will have
+              different results that may be worse or even better than the
+              estimation given here.
```
Am I seeing double?  Refactor?

innovaker reviewed (commented):
Nice work!  There's some really neat ideas in here and you've clearly put a lot of effort into it!

Below are notes that came to mind as I read through the code.  I appreciate it's still early days for this feature though.

- "Power profiler" implies it takes readings from your keyboard (to "profile" your power use), much as I believe the Nordic Power Profiler does?  Suggestions:
a) Battery Life Calculator
b) Power Usage Estimator/Guesstimator
c) Juice Projector/Predictor 😃 
or some combination/variation of the above words.

- Headers/licenses please!

- Are you intending to maintain this going forward?  I suspect ZMK is going to evolve somewhat this year which will probably invalidate the profiler's assumptions and data.  We can't expect firmware contributors to keep it in sync.

- It's safe to assume that the concept of "split" will be generalized to more than two nodes later this year.  It's probably worth keeping that in the back of your mind for design choices.

- Even with the big fat disclaimer, I still have concerns about introducing this feature without implementing a clear feedback loop for data from actual keyboards to verify and improve the model.  But I won't stand in the way of it going live.

- Does the page adapt to mobile devices well?  I'd imagine this particular page will be access from phones a lot.

Nicell commented on docs/src/components/power-estimate.js:
```diff
@@ -0,0 +1,242 @@
+
+const batVolt = lithiumIonAverageVoltage;
+
+const palette = [
+  "#bbdefb",
+  "#90caf9",
+  "#64b5f6",
+  "#42a5f5",
+  "#2196f3",
+  "#1e88e5",
+  "#1976d2",
+];
```
Due to how they're used, I feel they're better this way. Using them as classes would be pretty cumbersome

Nicell commented on docs/src/components/power-estimate.js:
```diff
@@ -0,0 +1,242 @@
+        <h3>... {splitType !== "standalone" ? "on " + splitType : ""}</h3>
+        <div className="powerEstimateBar">
+          <div
+            className="powerEstimateBarEl"
```
Yes, It means element. I can rename some that are a bit to vague.

Nicell commented on docs/src/components/power-estimate.js:
```diff
@@ -0,0 +1,242 @@
+                <div>
+                  {p.title} - {Math.round((p.usage / totalUsage) * 100) + "%"}
+                </div>
+                <div style={{ fontSize: "14px" }}>
```
Depends. I was a bit lazy here and should be using `rem`s instead.

Nicell commented on docs/src/data/power.js:
```diff
@@ -0,0 +1,55 @@
+  },
+};
+
+export const zmkBoards = {
```
Ideally we can have metadata files for each board. For now this will be held here.

Nicell commented:
> * "Power profiler" implies it takes readings from your keyboard (to "profile" your power use), much as I believe the Nordic Power Profiler does?  Suggestions:
>   EDIT: having now used Nordic's Online Power Profiler - I now appreciate where you got the name from.

Does this mean we can stick to this name? Your other suggestions just feel too long IMO. Thoughts?

> * Are you intending to maintain this going forward?  I suspect ZMK is going to evolve somewhat this year which will probably invalidate the assumptions and data.  We can't expect firmware contributors to keep it in sync.

Yes, I'll definitely plan to keep this maintained. Ideally I'll be updating the numbers on a schedule by retesting all of the values.

> * It's safe to assume that the concept of "split" will be generalized to more than two nodes later this year.  It's probably worth keeping that in the back of your mind for design choices.

This is a fair point. It should be pretty easy to extend this to handle that. 

> * Even with the big fat disclaimer, I still have concerns about introducing this feature without implementing a clear feedback loop for data from actual keyboards to verify and improve the model.  But I won't stand in the way of it going live.

I'm not sure if there's much more we can do. If it means anything, of the people I've asked, their battery life has fallen within the range besides a few exceptions (whom I think had some other power stuff going on they weren't accounting for. LEDs, OLEDs, etc).

> * Does the page adapt to mobile devices well?  I'd imagine this particular page will be accessed from phones a lot.
It does! Try it out.

Nicell commented on docs/src/components/power-estimate.js:
```diff
@@ -0,0 +1,240 @@
+
+  // Lithium ion self discharge
+  const lithiumDischarge =
+    ((parseInt(batterymAh) * 1000 * lithiumIonMonthlyDischargePercent) /
```
I've redone how these are all calculated so that the variables are much more direct and found the only real calculation I can pull out is the conversion from micro amps to milliwatt usage, which feels superfluous. Hopefully what I've updated to is much more readable (should only need to know ohms law and metric conversion now 😅).

Nicell commented:
I've addressed nearly everything from the reviews. I should note that the headers I've added are under the MIT license rather than the rest of the docs that has a creative commons license. I chose this, as I feel this work should be under MIT since it's not really directly documentation. If anyone has some opposition, I'm open to persuasion to use the other license instead.

joelspadin reviewed (approved):
There are still a few places in here with massive functions that could be split into smaller components, but this seems like a good start to me.

Nicell commented:
@innovaker @petejohanson, this PR is holding me up on a few projects. Any chance this could get a review soon? I have a feeling we could iterate on this forever, so hopefully we can focus on most iterations happening in future PRs.

markstos commented:
Cool project. As a USB keyboard user, I'm curious about the power draw of QMK vs ZMK.  I noticed there are USB power meters that could be added in-line between the laptop and the keyboard to measure the power usage over say, 24 hours:

https://www.amazon.com/Klein-Tools-ET920-Capacity-Resistance/dp/B07FMHHS43/

If someone had a keyboard with hot-swap sockets with both QMK and ZMK-programmed chips, it would be possible to check power draw for idling and "typical use" with otherwise identical PCBs and switches.  I'm curious if ZMKs low-power features translate to USB as well. (I suppose I should open a new thread for this...)

Nicell commented:
@markstos This power profiler is mostly targeted to handle power readings for batteries, but the results should be pretty close to what you'd get off USB. Of course, USB power pull isn't nearly as important as battery pull when wireless. Sure ZMK may idle at only 20uA higher than quiescent, but even 10mA will barely make a difference when connected over USB. A lot of this also comes down to the MCU of choice. The same hardware setup on QMK and ZMK *probably* won't have very different power usage.

petejohanson reviewed (approved):
I think this is a totally acceptable starting point for this. It's definitely used by folks already, and the numbers have been reported to be "on target", which combined w/ the disclaimer, is enough for me. Thanks @Nicell !
